src/beamlime/dashboard/workflow_controller.py

```diff
@@ -0,0 +1,249 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2025 Scipp contributors (https://github.com/scipp)
+"""
+Workflow controller implementation backed by a config service.
+"""
+
+import logging
+from collections.abc import Callable
+from typing import Any
+
+from beamlime.config.workflow_spec import (
+    PersistentWorkflowConfig,
+    WorkflowConfig,
+    WorkflowId,
+    WorkflowSpec,
+    WorkflowSpecs,
+    WorkflowStatus,
+    WorkflowStatusType,
+)
+
+from .workflow_config_service import ConfigServiceAdapter, WorkflowConfigService
+from .workflow_controller_base import WorkflowControllerBase
+
+
+class WorkflowController(WorkflowControllerBase):
```

Comment:
Why does this inherit from `WorkflowControllerBase`? The latter is a protocol but `WorkflowController` looks like a full implementation, not a protocol.

Why does the protocol exist in the first place? There seems to be only one implementation.